osf/external/gravy_valet/request_helpers.py

```diff
@@ -0,0 +1,221 @@
+        params={'include': ACCOUNT_EXTERNAL_SERVICE_PATH},
+    )
+
+
+def iterate_addons_for_resource(requested_resource, requesting_user):  # -> typing.Iterator[JSONAPIResultEntry]
+    '''Returns an iterator of JSONAPIResultEntires representing all of the ConfiguredStorageAddons for a resource.'''
+    resource_result = get_gv_result(
+        endpoint_url=RESOURCE_FILTER_ENDPOINT.format(uri=requested_resource.get_semantic_iri()),
+        requesting_user=requesting_user,
+        requested_resource=requested_resource,
+    )
+    if not resource_result:
+        return None
+    yield from iterate_gv_results(
+        endpoint_url=resource_result.get_related_link('configured_storage_addons'),
+        requesting_user=requesting_user,
+        requested_resource=requested_resource,
+        params={'include': ADDON_EXTERNAL_SERVICE_PATH}
+    )
+
+
+def get_waterbutler_config(gv_addon_pk, requested_resource, requesting_user):  # -> JSONAPIResultEntry
+    return get_gv_result(
+        endpoint=WB_CONFIG_ENDPOINT.format(pk=gv_addon_pk),
+        requesting_user=requesting_user,
+        requested_resource=requested_resource
+    )
+
+
+def get_gv_result(**kwargs):  # -> JSONAPIResultEntry
+    '''Processes the result of a request to a GravyValet detail endpoint into a JSONAPIResultEntry.
+
+    kwargs must match _make_gv_request
+    '''
+    response = _make_gv_request(**kwargs)
+    response_json = response.json()
+    if not response['data']:
+        return None
+    included_entities_lookup = _format_included_entities(response_json.get('included', []))
+    return JSONAPIResultEntry(response_json['data'], included_entities_lookup)
+
+
+def iterate_gv_results(**kwargs):  # -> typing.Iterator[JSONAPIResultEntry]
+    '''Processes the result of a request to GravyValet list endpoint into a generator of JSONAPIResultEntires.
+
+    kwargs must match _make_gv_request
+    '''
+
+    response_json = _make_gv_request(**kwargs).json()
+    if not response_json['data']:
+        return  # empty iterator
+    included_entities_lookup = _format_included_entities(response_json.get('included', []))
+    yield from (JSONAPIResultEntry(entry, included_entities_lookup) for entry in response_json['data'])
+
+
+def _make_gv_request(
+    endpoint_url: str,
+    requesting_user,
+    requested_resource=None,
+    request_method='GET',
+    params: dict = None
+):
+    '''Generates HMAC-Signed auth headers and makes a request to GravyValet, returning the result.'''
+    full_url = urlunparse(urlparse(endpoint_url)._replace(query=urlencode(params)))
+    auth_headers = auth_helpers.make_gravy_valet_hmac_headers(
+        request_url=full_url,
+        request_method=request_method,
+        additional_headers=auth_helpers._make_permissions_headers(
+            requesting_user=requesting_user,
+            requested_resource=requested_resource
+        )
+    )
+    response = requests.get(full_url, headers=auth_headers, params=params)
```

Comment:
actually meant to use `endpoint_url` here but only because of some nebulous feeling that it's more "correct" to set query parameters that way than by creating the full url. Just needed the ful url for making headers, which is maybe a smell that that should change...